### PR TITLE
IGNITE-15329 Atomics should be repairable by Read Repair

### PR DESCRIPTION
**Pull request by anton-vinogradov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [x] There is a single JIRA ticket related to the pull request. 
- [x] The web-link to the pull request is attached to the JIRA ticket.
- [x] The JIRA ticket has the _Patch Available_ state.
- [x] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [x] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/consistency/IgniteAtomicConsistencyViolationException.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    /**
+     *  @param fixedMap Fixed map.
+     * @param primaryMap Primary map.
+     * @param callback Fixed callback.
+     */
+    public IgniteAtomicConsistencyViolationException(
+        Map<KeyCacheObject, EntryGetResult> fixedMap,
+        Map<KeyCacheObject, EntryGetResult> primaryMap,
+        Consumer<Map<KeyCacheObject, EntryGetResult>> callback) {
+        super(fixedMap.keySet());
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/consistency/IgniteAtomicConsistencyViolationException.java:51]: Empty line required

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/consistency/IgniteAtomicConsistencyViolationException.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    private final Map<KeyCacheObject, EntryGetResult> fixedMap;
+
+    /** Primary map. */
+    private final Map<KeyCacheObject, EntryGetResult> primaryMap;
+
+    /** On fixed callback. */
+    private final Consumer<Map<KeyCacheObject, EntryGetResult>> callback;
+
+    /**
+     *  @param fixedMap Fixed map.
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/consistency/IgniteAtomicConsistencyViolationException.java:43]: Typo: extra space

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/ReadRepairDataGenerator.java [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+        this.serverNodesCnt = serverNodesCnt;
+        this.backupsCnt = backupsCnt;
+    }
+
+    /**
+     *
+     */
+    public void generate(
+        Ignite initiator,
+        Integer cnt,
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/ReadRepairDataGenerator.java:127]: `cnt` can be int.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/ReadRepairDataGenerator.java [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+    private final BiFunction<Integer, String, List<Ignite>> backupNodes;
+
+    /** Server nodes count. */
+    private final Supplier<Integer> serverNodesCnt;
+
+    /** Backups count. */
+    private final Supplier<Integer> backupsCnt;
+
+    /**
+     * @param iterableKey Iterable key.
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/ReadRepairDataGenerator.java:94]: Looks like `iterableKey` can be incapsulated inside generator.

**Comment by pkgonan** [comment]
Hello, is this work going to be applied to GridGain?

**Comment by nizhikov** [comment]
@pkgonan I think you should ask Grid Gain support or employee.

This patch is open source and CAN be applied to any Ignite fork.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/AbstractReadRepairTest.java [hunk 5]
```diff
@@ -237,429 +256,74 @@ else if (e.irreparableKeys().contains(key) || (e.repairableKeys() != null && e.r
...
-        }
+    protected void generateAndCheck(
+        Ignite initiator,
+        int cnt,
+        boolean raw,
+        boolean async,
+        boolean misses,
+        boolean nulls,
+        boolean binary,
+        ReadRepairStrategy strategy,
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/AbstractReadRepairTest.java:316]: strategy is always null and can be omitted.

Can we create special `generate` method for the cases when `strategy` is null?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/AbstractFullSetReadRepairTest.java [hunk 6]
```diff
@@ -40,80 +42,78 @@ public abstract class AbstractFullSetReadRepairTest extends AbstractReadRepairTe
...
-    protected static final Consumer<ReadRepairData> GET_CHECK_AND_FIX = (data) -> {
-        IgniteCache<Integer, Integer> cache = data.cache;
+    protected static final Consumer<ReadRepairData> GET_CHECK_AND_REPAIR = (data) -> {
+        IgniteCache<Integer, Object> cache = data.cache;
+
         Set<Integer> keys = data.data.keySet();
         boolean raw = data.raw;
         boolean async = data.async;
+        boolean binary = data.binary;
         ReadRepairStrategy strategy = data.strategy;
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/AbstractFullSetReadRepairTest.java:52]: Let's inline `keys`, `raw`, `async, `strtegy`.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/AbstractFullSetReadRepairTest.java [hunk 7]
```diff
@@ -40,80 +42,78 @@ public abstract class AbstractFullSetReadRepairTest extends AbstractReadRepairTe
...
+            assertEquals(repaired, entry.getValue());
         }
     };
 
     /**
      *
      */
     protected static final Consumer<ReadRepairData> GET_NULL = (data) -> {
-        IgniteCache<Integer, Integer> cache = data.cache;
+        IgniteCache<Integer, Object> cache = data.binary ? data.cache.withKeepBinary() : data.cache;
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/AbstractFullSetReadRepairTest.java:115]: Let's inline `raw`, `async`, `strategy`, `keys`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheAdapter.java [hunk 8]
```diff
@@ -5134,6 +5139,105 @@ private IgniteInternalFuture<Void> repairAsync(
...
+            this.primVal = primVal;
+            this.primVer = primVer;
+        }
+
+        /** {@inheritDoc} */
+        @Override public Boolean process(MutableEntry<K, V> entry, Object... arguments) throws EntryProcessorException {
+            V entryVal = entry.getValue();
+
+            try {
+                if ((primVal == null && entryVal == null) || // Still null at primary.
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheAdapter.java:5220]:
```
(primVal == null && entryVal == null) || // Still null at primary.
                    // No updates since consistency violation has been found.
                    ((primVal != null && primVal.equals(entryVal)
```
Can be replaced with `Objects.equals`.

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/consistency/ConsistencyCommand.java [hunk 9]
```diff
@@ -52,43 +57,64 @@ public class ConsistencyCommand extends AbstractCommand<Object> {
...
-                BROADCAST_UUID,
-                clientCfg
-            );
-
-            if (res.cancelled()) {
-                sb.append("Operation execution cancelled.\n\n");
-
-                failed = true;
-            }
+            Set<UUID> nodeIds = parallel ?
```
Comment (nizhikov) [inline modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/consistency/ConsistencyCommand.java:82]: Code style.

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyRepairCorrectnessTransactionalTest.java [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+@RunWith(Parameterized.class)
+@WithSystemProperty(key = IGNITE_ENABLE_EXPERIMENTAL_COMMAND, value = "true")
+public class GridCommandHandlerConsistencyRepairCorrectnessTransactionalTest extends GridCommandHandlerAbstractTest {
+    /** Test parameters. */
+    @Parameterized.Parameters(name = "misses={0}, nulls={1}, strategy={2}, parallel={3}")
+    public static Iterable<Object[]> parameters() {
+        List<Object[]> res = new ArrayList<>();
+
+        for (boolean misses : new boolean[] {false, true}) {
+            for (boolean nulls : new boolean[] {false, true})
```
Comment (nizhikov) [inline modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyRepairCorrectnessTransactionalTest.java:58]: Code style.

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyTest.java [hunk 11]
```diff
@@ -279,42 +290,39 @@ private void readRepair(AtomicInteger brokenParts, String cacheName, Integer fix
...
-                else
+                if (brokenParts.get() > 0)
                     assertContains(log, testOut.toString(),
-                        "conflict partitions has been found: [counterConflicts=0, hashConflicts=" + brokenParts); // Nothing fixed.
+                        "conflict partitions has been found: [counterConflicts=0, hashConflicts=" + brokenParts);
+                else
+                    assertContains(log, testOut.toString(), "no conflicts have been found");
+            }
+            else
+                assertContains(log, testOut.toString(),
```
Comment (nizhikov) [inline modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyTest.java:307]: Code style.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheAdapter.java [hunk 12]
```diff
@@ -5068,19 +5071,21 @@ private <R> IgniteInternalFuture<R> getWithRepairAsync(
...
     private IgniteInternalFuture<Void> repairAsync(
-        Collection<KeyCacheObject> keys,
+        IgniteConsistencyViolationException ex,
         final CacheOperationContext opCtx,
         boolean skipVals) {
         GridCompoundReadRepairFuture fut = new GridCompoundReadRepairFuture();
 
-        for (KeyCacheObject key : keys)
-            fut.add(repairAsync(key, opCtx, skipVals));
+        for (KeyCacheObject key : ex.keys())
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheAdapter.java:5085]: Multiline for. Braces required.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/consistency/GridNearReadRepairAbstractFuture.java [hunk 13]
```diff
@@ -71,7 +75,7 @@ public abstract class GridNearReadRepairAbstractFuture extends GridFutureAdapter
     protected final Map<ClusterNode, GridPartitionedGetFuture<KeyCacheObject, EntryGetResult>> futs;
 
     /** Context. */
-    protected final GridCacheContext<KeyCacheObject, EntryGetResult> ctx;
+    protected final GridCacheContext<?, ?> ctx;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/consistency/GridNearReadRepairAbstractFuture.java:78]: Why this change?
Can be undone.

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/consistency/ConsistencyCommand.java (hunk 9 above)
Comment (anton-vinogradov) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/consistency/ConsistencyCommand.java:82]: What should be changed?

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyRepairCorrectnessTransactionalTest.java (hunk 10 above)
Comment (anton-vinogradov) [reply modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyRepairCorrectnessTransactionalTest.java:58]: What should be changed?

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyTest.java [hunk 14]
```diff
@@ -148,9 +151,20 @@ private void testAtomicAndTx(boolean incVal) throws Exception {
...
         injectTestSystemOut();
 
-        LogListener lsnrUnmaskedKey = matches("Key: 0 (cache: ").build(); // Unmasked key.
-        LogListener lsnrMaskedKey = matches("Key: [HIDDEN_KEY#").build(); // Masked key.
-        LogListener lsnrMaskedVal = matches("Value: [HIDDEN_VALUE#").build(); // Masked value.
+        int repairsPerEntry = repairsPerEntry();
+
+        int copies = BACKUPS + 1;
+
+        int timesMultiplicator = repairsPerEntry == 0 ?
```
[the remark this answers is not in the document]
Comment (anton-vinogradov) [reply modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyTest.java:158]: What should be changed?

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyTest.java (hunk 11 above)
Comment (anton-vinogradov) [reply modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyTest.java:307]: What should be changed?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/consistency/GridNearReadRepairAbstractFuture.java (hunk 13 above)
Comment (anton-vinogradov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/consistency/GridNearReadRepairAbstractFuture.java:78]: This will cause warnings.

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/consistency/ConsistencyCommand.java [hunk 15]
```diff
@@ -184,6 +217,11 @@ public class ConsistencyCommand extends AbstractCommand<Object> {
             if (strategy == null)
                 throw new IllegalArgumentException("Strategy argument missed.");
 
+            // see https://issues.apache.org/jira/browse/IGNITE-15316
+            if (parallel && strategy != ReadRepairStrategy.CHECK_ONLY)
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/consistency/ConsistencyCommand.java:221]: multiline if require braces.

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyRepairCorrectnessTransactionalTest.java (hunk 10 above)
Comment (nizhikov) [reply modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyRepairCorrectnessTransactionalTest.java:58]: multiline for require braces.

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyTest.java (hunk 11 above)
Comment (nizhikov) [reply modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyTest.java:307]: multiline if require braces.

**Review by nizhikov — APPROVED** [review]
